File: abstract_factory/cpp/abstract_factory.cc
#include <iostream>
#include "abstract_factory/cpp/abstract_factory.h"

void XmlOrderMainDAO::save_order_main() {
    std::cout << "xml save main" << '\n';
}

void RdbOrderMainDAO::save_order_main() {
    std::cout << "rdb save main" << '\n';
}

void XmlOrderDetailDAO::save_order_detail() {
    std::cout << "xml save detail" << '\n';
}

void RdbOrderDetailDAO::save_order_detail() {
    std::cout << "rdb save detail" << '\n';
}

OrderMainDAO* XmlDAOFactory::create_order_main_dao() {
    return new XmlOrderMainDAO();
}

OrderDetailDAO* XmlDAOFactory::create_order_detail_dao() {
    return new XmlOrderDetailDAO();
}

OrderMainDAO* RdbDAOFactory::create_order_main_dao() {
    return new RdbOrderMainDAO();
}

OrderDetailDAO* RdbDAOFactory::create_order_detail_dao() {
    return new RdbOrderDetailDAO();
}

File: abstract_factory/abstract_factory.go
package abstractfactory

type OrderMainDAO interface {
	SaveOrderMain() string
}

type OrderDetailDAO interface {
	SaveOrderDetail() string
}

type DAOFactory interface {
	CreateOrderMainDAO() OrderMainDAO
	CreateOrderDetailDAO() OrderDetailDAO
}

type RDBOrderMainDAO struct{}

func (RDBOrderMainDAO) SaveOrderMain() string {
	return "rdb main save"
}

type RDBOrderDetailDAO struct{}

func (RDBOrderDetailDAO) SaveOrderDetail() string {
	return "rdb detail save"
}

type XMLOrderMainDAO struct{}

func (XMLOrderMainDAO) SaveOrderMain() string {
	return "xml main save"
}

type XMLOrderDetailDAO struct{}

func (XMLOrderDetailDAO) SaveOrderDetail() string {
	return "xml detail save"
}

type RDBDAOFactory struct{}

func (RDBDAOFactory) CreateOrderMainDAO() OrderMainDAO {
	return RDBOrderMainDAO{}
}

func (RDBDAOFactory) CreateOrderDetailDAO() OrderDetailDAO {
	return RDBOrderDetailDAO{}
}

type XMLDAOFactory struct{}

func (XMLDAOFactory) CreateOrderMainDAO() OrderMainDAO {
	return XMLOrderMainDAO{}
}

func (XMLDAOFactory) CreateOrderDetailDAO() OrderDetailDAO {
	return XMLOrderDetailDAO{}
}
